File: core/compiler/templates/objc/UMCore.m
#import "UMCore.h"

@implementation UMCore
- (instancetype)initWithJson:(NSString *)json {
	return [super init];
}

- (NSString *)encodeToJson {
	return @"";
}
@end
{% for namespace in namespaces %}
{% for object in namespace.objects %}

@implementation UMCore{{ namespace.name }}{{ object.name }}
- (instancetype)init {
	if (self = [super init]) {
		{% for attribute in object.attributes -%}
		{%- if attribute|um_attribute_is("STRING") -%}
		{%- if attribute.value|um_value_is("object") -%}
		_{{ attribute.name }} = kUMCore{{ attribute.value.namespace }}{{ attribute.value.object }}{{ attribute.value.constant }};
		{%- else -%}
		_{{ attribute.name }} = @"{{ attribute.value }}";
		{%- endif -%}
		{%- elseif attribute|um_attribute_is("FLOAT") || attribute|um_attribute_is("BOOLEAN") || attribute|um_attribute_is("INTEGER") -%}
		{%- if attribute.value|um_value_is("object") -%}
		_{{ attribute.name }} = kUMCore{{ attribute.value.namespace }}{{ attribute.value.object }}{{ attribute.value.constant }};
		{%- else -%}
		_{{ attribute.name }} = {{ attribute.value }};
		{%- endif -%}
		{%- elseif attribute|um_attribute_is("OBJECT") -%}
		_{{ attribute.name }} = [UMCore{{ attribute.value.namespace }}{{ attribute.value.object }} new];
		{%- elseif attribute|um_attribute_is("ARRAY") -%}
		_{{ attribute.name }} = [NSArray<UMCore{{ attribute.value.namespace }}{{ attribute.value.object }} *> array];
		{%- endif -%}
		{%- endfor %}
	}
	return self;
}

- (instancetype)initWithJson:(NSDictionary *)json {
	if (self = [self init]) {
		id attribute = nil;

		{% for attribute in object.attributes -%}
		attribute = [json objectForKey:@"{{ attribute.name }}"];
		if (attribute) {
			{% if attribute|um_attribute_is("STRING") -%}
			_{{ attribute.name }} = attribute;
			{%- elseif attribute|um_attribute_is("BOOLEAN") -%}
			_{{ attribute.name }} = ((NSNumber *)attribute).boolValue;
			{%- elseif attribute|um_attribute_is("FLOAT") -%}
			_{{ attribute.name }} = ((NSNumber *)attribute).doubleValue;
			{%- elseif attribute|um_attribute_is("INTEGER") -%}
			_{{ attribute.name }} = ((NSNumber *)attribute).integerValue;
			{%- elseif attribute|um_attribute_is("OBJECT") -%}
			_{{ attribute.name }} = [[UMCore{{ attribute.value.namespace }}{{ attribute.value.object }} alloc] initWithJson:attribute];
			{%- elseif attribute|um_attribute_is("ARRAY") -%}
			NSMutableArray<UMCore{{ attribute.value.namespace }}{{ attribute.value.object }} *> *array = [NSMutableArray<UMCore{{ attribute.value.namespace }}{{ attribute.value.object }} *> array];
			for(NSDictionary *item in attribute) {
				[array addObject:[[UMCore{{ attribute.value.namespace }}{{ attribute.value.object }} alloc] initWithJson:item]];
			}
			_{{ attribute.name }} = array;
			{%- endif %}
		}
		else {
			{% if attribute|um_attribute_is("STRING") -%}
			{%- if attribute.value|um_value_is("object") -%}
			_{{ attribute.name }} = kUMCore{{ attribute.value.namespace }}{{ attribute.value.object }}{{ attribute.value.constant }};
			{%- else -%}
			_{{ attribute.name }} = @"{{ attribute.value }}";
			{%- endif -%}
			{%- elseif attribute|um_attribute_is("FLOAT") || attribute|um_attribute_is("BOOLEAN") || attribute|um_attribute_is("INTEGER") -%}
			{%- if attribute.value|um_value_is("object") -%}
			_{{ attribute.name }} = kUMCore{{ attribute.value.namespace }}{{ attribute.value.object }}{{ attribute.value.constant }};
			{%- else -%}
			_{{ attribute.name }} = {{ attribute.value }};
			{%- endif -%}
			{%- elseif attribute|um_attribute_is("OBJECT") -%}
			_{{ attribute.name }} = [UMCore{{ attribute.value.namespace }}{{ attribute.value.object }} new];
			{%- elseif attribute|um_attribute_is("ARRAY") -%}
			_{{ attribute.name }} = [NSArray<UMCore{{ attribute.value.namespace }}{{ attribute.value.object }} *> array];
			{%- endif -%}
		}
	{%- endfor %}
	}
	return self;
}

- (NSDictionary *)encodeToJson {
	NSMutableDictionary *json = [NSMutableDictionary dictionary];

	{% for attribute in object.attributes -%}
	{%- if attribute|um_attribute_is("STRING") -%}
	[json setObject:_{{ attribute.name }} forKey:@"{{ attribute.name }}"];
	{%- elseif attribute|um_attribute_is("BOOLEAN") || attribute|um_attribute_is("FLOAT") || attribute|um_attribute_is("INTEGER") -%}
	[json setObject:@(_{{ attribute.name }}) forKey:@"{{ attribute.name }}"];
	{%- elseif attribute|um_attribute_is("OBJECT") -%}
	[json setObject:[_{{ attribute.name }} encodeToJson] forKey:@"{{ attribute.name }}"];
	{%- elseif attribute|um_attribute_is("ARRAY") -%}
	{
		NSMutableArray *array = [NSMutableArray array];
		for(UMCore{{ attribute.value.namespace}}{{ attribute.value.object }} *item in _{{ attribute.name }}) {
			[array addObject:[item encodeToJson]];
		}
		[json setObject:array forKey:@"{{ attribute.name }}"];
	}
	{%- endif -%}
	{%- endfor %}
	return json;
}

{% for attribute in object.attributes %}
- (void)set{{ attribute.name|ucfirst }}:({% if attribute|um_attribute_is("STRING") %}NSString *{% elseif attribute|um_attribute_is("INTEGER") %}NSInteger {% elseif attribute|um_attribute_is("FLOAT") %}double {% elseif attribute|um_attribute_is("BOOLEAN") %}BOOL {% elseif attribute|um_attribute_is("OBJECT") || attribute|um_attribute_is("ARRAY") %}UMCore{{attribute.value.namespace}}{{attribute.value.object}} *{% endif %}){{ attribute.name }} {
	{% if attribute.filter -%}
	{%- if attribute.filter|um_filter_is("object") -%}
	if (({{ attribute.name }} <{% if !attribute.filter.include.lower %}={% endif %} {{ attribute.filter.range.lower }})
		|| ({{ attribute.name}} >{% if !$attribute.filter.include.upper %}={% endif %} {{ attribute.filter.range.upper }})) {
		[NSException raise:@"UMCore" format:@"out of range"];
		return;
	}
	{%- else -%}
	NSRegularExpression *regex = [NSRegularExpression regularExpressionWithPattern:@"{{ attribute.filter }}" options:0 error:nil];
	if (![regex numberOfMatchesInString:{{ attribute.name }} options:0 range:NSMakeRange(0, {{ attribute.name }}.length)]) {
		[NSException raise:@"UMCore" format:@"mismatch regex"];
		return;
	}
	{%- endif -%}
	{%- endif %}
	_{{ attribute.name }} = {{ attribute.name }};
}
{%- endfor %}
@end
{%- endfor -%}
{%- endfor %}

File: core/compiler/templates/java/UMCore.java
package com.vanchu.lib.core.um;
import java.util.LinkedList;
import java.util.regex.Pattern;
import org.json.JSONObject;
import org.json.JSONArray;

public class UMCore {
	{% for namespace in namespaces -%}
	public static final class {{ namespace.name }} {
		{% for object in namespace.objects %}
		public static final class {{ object.name }} {
			{% for constant in object.constants -%}
			public static final {% if constant|um_constant_is("STRING") %}String{% elseif constant|um_constant_is("INTEGER") %}int{% elseif constant|um_constant_is("FLOAT") %}double{% elseif constant|um_constant_is("BOOLEAN") %}Boolean{% endif %} {{ constant.name }} = {% if constant|um_constant_is("STRING") %}"{{ constant.value }}"{% else %}{{ constant.value }}{% endif %};
			{%- endfor -%}
			{% for attribute in object.attributes -%}
			private {% if attribute|um_attribute_is("STRING") %}String{% elseif attribute|um_attribute_is("INTEGER") %}int{% elseif attribute|um_attribute_is("FLOAT") %}double{% elseif attribute|um_attribute_is("BOOLEAN") %}Boolean{% elseif attribute|um_attribute_is("OBJECT") %}{{ attribute.value.namespace }}.{{ attribute.value.object }}{% elseif attribute|um_attribute_is("ARRAY") %}LinkedList<{{ attribute.value.namespace }}.{{ attribute.value.object }}>{% endif %} _{{ attribute.name }};
			{%- endfor %}

			public {{ object.name }}() {
				{% for attribute in object.attributes -%}
				this._{{ attribute.name }} = <{$attribute|um_java_attribute_get_default_value}>;
				{%- if attribute|um_attribute_is("STRING") -%}
				{%- if attribute.value|um_value_is("object") -%}
				this._{{ attribute.name }} = {{ attribute.value.namespace }}.{{ attribute.value.object }}.{{ attribute.value.constant }};
				{%- else -%}
				this._{{ attribute.name }} = "{{ attribute.value }}";
				{%- endif -%}
				{%- elseif attribute|um_attribute_is("INTEGER") || attribute|um_attribute_is("FLOAT") || attribute|um_attribute_is("BOOLEAN") -%}
				{%- if attribute.value|um_value_is("object") -%}
				this._{{ attribute.name }} = {{ attribute.value.namespace }}.{{ attribute.value.object }}.{{ attribute.value.constant }};
				{%- else -%}
				this._{{ attribute.name }} = {{ attribute.value }};
				{%- endif -%}
				{%- elseif attribute|um_attribute_is("OBJECT") -%}
				this._{{ attribute.name }} = new {{ attribute.value.namespace }}.{{ attribute.value.object }}();
				{%- elseif attribute|um_attribute_is("ARRAY") -%}
				this._{{ attribute.name }} = new LinkedList<{{ attribute.value.namespace }}.{{ attribute.value.object }}>();
				{%- endif -%}
				{%- endfor %}
			}

			public {{ object.name }}(JSONObject json) {
				{% for attribute in object.attributes -%}
				if (json.has("{{ attribute.name }}")) {
					{% if attribute|um_attribute_is("STRING") -%}
					this._{{ attribute.name }} = json.getString("{{ attribute.name }}");
					{%- elseif attribute|um_attribute_is("BOOLEAN") -%}
					this._{{ attribute.name }} = json.getBoolean("{{ attribute.name }}");
					{%- elseif attribute|um_attribute_is("FLOAT") -%}
					this._{{ attribute.name }} = json.getDouble("{{ attribute.name }}");
					{%- elseif attribute|um_attribute_is("INTEGER") -%}
					this._{{ attribute.name }} = json.getInt("{{ attribute.name }}");
					{%- elseif attribute|um_attribute_is("OBJECT") -%}
					this._{{ attribute.name }} = new {{ attribute.value.namespace }}.{{ attribute.value.object }}(json.getJSONObject("{{ attribute.name }}"));
					{%- elseif attribute|um_attribute_is("ARRAY") -%}
					this._{{ attribute.name }} = new LinkedList<{{ attribute.value.namespace }}.{{ attribute.value.object }}>();
					for(Object item : json.getJSONArray("{{ attribute.name }}")) {
						this._{{ attribute.name }}.add(new {{ attribute.value.namespace }}.{{ attribute.value.object }}((JSONObject)item));
					}
					{%- endif %}
				}
				else {
					this._{{ attribute.name }} = <{$attribute|um_java_attribute_get_default_value}>;
					{%- if attribute|um_attribute_is("STRING") -%}
					{%- if attribute.value|um_value_is("object") -%}
					this._{{ attribute.name }} = {{ attribute.value.namespace }}.{{ attribute.value.object }}.{{ attribute.value.constant }};
					{%- else -%}
					this._{{ attribute.name }} = "{{ attribute.value }}";
					{%- endif -%}
					{%- elseif attribute|um_attribute_is("INTEGER") || attribute|um_attribute_is("FLOAT") || attribute|um_attribute_is("BOOLEAN") -%}
					{%- if attribute.value|um_value_is("object") -%}
					this._{{ attribute.name }} = {{ attribute.value.namespace }}.{{ attribute.value.object }}.{{ attribute.value.constant }};
					{%- else -%}
					this._{{ attribute.name }} = {{ attribute.value }};
					{%- endif -%}
					{%- elseif attribute|um_attribute_is("OBJECT") -%}
					this._{{ attribute.name }} = new {{ attribute.value.namespace }}.{{ attribute.value.object }}();
					{%- elseif attribute|um_attribute_is("ARRAY") -%}
					this._{{ attribute.name }} = new LinkedList<{{ attribute.value.namespace }}.{{ attribute.value.object }}>();
					{%- endif%}
				}
				{%- endfor %}
			}

			public JSONObject encodeToJson() {
				JSONObject json = new JSONObject();
				{% for attribute in object.attributes -%}
				{%- if attribute|um_attribute_is("STRING") || attribute|um_attribute_is("INTEGER") || attribute|um_attribute_is("FLOAT") || attribute|um_attribute_is("BOOLEAN") -%}
				json.put("{{ attribute.name }}", this._{{ attribute.name }});
				{%- elseif attribute|um_attribute_is("OBJECT") -%}
				json.put("{{ attribute.name }}", this._{{ attribute.name }}.encodeToJson());
				{%- elseif attribute|um_attribute_is("ARRAY") -%}
				{
					JSONArray array = new JSONArray();
					for({{ attribute.value.namespace }}.{{ attribute.value.object }} item : this._{{ attribute.name }}) {
						array.put(item.encodeToJson());
					}
					json.put("{{ attribute.name }}", array);
				}
				{%- endif -%}
				{%- endfor %}
				return json;
			}

			{% for attribute in object.attributes %}
			public {% if attribute|um_attribute_is("STRING") %}String{% elseif attribute|um_attribute_is("INTEGER") %}int{% elseif attribute|um_attribute_is("FLOAT") %}double{% elseif attribute|um_attribute_is("BOOLEAN") %}Boolean{% elseif attribute|um_attribute_is("OBJECT") %}{{ attribute.value.namespace }}.{{ attribute.value.object }}{% elseif attribute|um_attribute_is("ARRAY") %}LinkedList<{{ attribute.value.namespace }}.{{ attribute.value.object }}>{% endif %} get{{ attribute.name|ucfirst }}() {
				return this._{{ attribute.name }};
			}

			public void set{{ attribute.name|ucfirst }}({% if attribute|um_attribute_is("STRING") %}String{% elseif attribute|um_attribute_is("INTEGER") %}int{% elseif attribute|um_attribute_is("FLOAT") %}double{% elseif attribute|um_attribute_is("BOOLEAN") %}Boolean{% elseif attribute|um_attribute_is("OBJECT") %}{{ attribute.value.namespace }}.{{ attribute.value.object }}{% elseif attribute|um_attribute_is("ARRAY") %}LinkedList<{{ attribute.value.namespace }}.{{ attribute.value.object }}>{% endif %} {{ attribute.name }}) throws Exception {
				{% if attribute.filter -%}
				{%- if attribute.filter|um_filter_is("object") -%}
				if (({{ attribute.name }} <{% if !attribute.filter.include.lower %}={% endif %} {{ attribute.filter.range.lower }}) || ({{ attribute.name }} >{% if !attribute.filter.include.upper %}={% endif %} {{ attribute.filter.range.upper }})) {
					throw new Exception("out of range");
				}
				{%- else -%}
				if(!Pattern.compile("{{ attribute.filter }}").matcher({{ attribute.name }}).matches()) {
					throw new Exception("mismatch regex");
				}
				{%- endif -%}
				{%- endif %}
				this._{{ attribute.name }} = {{ attribute.name }};
			}
			{%- endfor %}
		}
		{%- endfor %}
	}
	{%- endfor %}
}
